packages/matter.js/src/device/Endpoint.ts

```diff
@@ -169,33 +203,39 @@ export class Endpoint {
         return this.childEndpoints;
     }
 
-    findHighestEndpointId(): number {
-        let highestEndpointId = 0;
-        this.childEndpoints.forEach(endpoint => {
-            const endpointId = endpoint.findHighestEndpointId();
-            if (endpointId > highestEndpointId) {
-                highestEndpointId = endpointId;
-            }
-        });
-        if (this.id !== undefined && this.id > highestEndpointId) {
-            highestEndpointId = this.id;
+    protected removeChildEndpoint(endpoint: Endpoint): void {
+        const index = this.childEndpoints.indexOf(endpoint);
+        if (index === -1) {
+            throw new Error(`Provided endpoint for deletion does not exist as child endpoint.`);
         }
-        return highestEndpointId;
+        this.childEndpoints.splice(index, 1);
+        endpoint.clearStructureChangedCallback(); // remove
+        this.structureChangedCallback(); // Inform parent about structure change
     }
 
-    ensureEndpointIds(nextEndpointId: number): number {
-        if (this.id === undefined) {
-            this.id = nextEndpointId++;
+    determineUniqueID(): string {
+        // if the options in constructor contained a custom uniqueId, use this
+        if (this.uniqueId !== undefined) {
+            return `customUniqueId_${this.uniqueId}`;
         }
-        if (this.childEndpoints.length) {
-            this.childEndpoints.forEach(endpoint => {
-                nextEndpointId = endpoint.ensureEndpointIds(nextEndpointId);
-            });
+        // Else we check if we have a basic information cluster or bridged device basic information cluster and
+        // use the uniqueId or serial number, if provided
+        const basicInformationCluster = this.getClusterServer(BasicInformationCluster) ?? this.getClusterServer(BridgedDeviceBasicInformationCluster);
+        if (basicInformationCluster !== undefined) {
+            const uniqueId = basicInformationCluster.getUniqueIdAttribute?.();
+            if (uniqueId !== undefined) {
+                return `basicInfoUniqueId_${uniqueId}`;
+            }
+            const serialNumber = basicInformationCluster.getSerialNumberAttribute?.();
+            if (serialNumber !== undefined) {
+                return `basicInfoSerialNumber_${serialNumber}`;
+            }
         }
-        return nextEndpointId;
+        // Else we have no unique id
+        return "";
```

Comment:
Why use `""` rather than `undefined` to indicate absence of an ID?

Reply:
ok